Invalid parsing of wildcard closure parameter

<!--
Thank you for filing a bug report! 🐛 Please provide a short summary of the bug,
along with any information you feel relevant to replicating the bug.
-->

I tried this code:

```rust
// { dg-additional-options "-fsyntax-only" }

pub fn encode_normal<T: RawFloat>(x: Unpacked) -> T {
    debug_assert!(
        T::MIN_SIG <= x.sig && x.sig <= T::MAX_SIG,
        "encode_normal: significand not normalized"
    );
    // Remove the hidden bit
    let sig_enc = x.sig & !(1 << T::EXPLICIT_SIG_BITS);
    // Adjust the exponent for exponent bias and mantissa shift
    let k_enc = x.k + T::MAX_EXP + T::EXPLICIT_SIG_BITS as i16;
    debug_assert!(k_enc != 0 && k_enc < T::MAX_ENCODED_EXP, "encode_normal: exponent out of range");
    // Leave sign bit at 0 ("+"), our numbers are all positive
    let bits = (k_enc as u64) << T::EXPLICIT_SIG_BITS | sig_enc;
    T::from_bits(bits.try_into().unwrap_or_else(|_| unreachable!()))
}
```

similar testcase:
```rust
fn foo() {
    let a = |_| 15;
}
```

I expected to see this happen: no error

Instead, this happened: 
```rust
../../rust/library/core/src/num/dec2flt/rawfp.rs:295:67: error: expecting ‘|’ but ‘)’ found
  295 |     T::from_bits(bits.try_into().unwrap_or_else(|_| unreachable!()))
      |                                                                   ^
../../rust/library/core/src/num/dec2flt/rawfp.rs:295:49: error: failed to parse method param in method call
  295 |     T::from_bits(bits.try_into().unwrap_or_else(|_| unreachable!()))
      |                                                 ^
../../rust/library/core/src/num/dec2flt/rawfp.rs:295:18: error: failed to parse expression in struct (or enum) expression tuple
  295 |     T::from_bits(bits.try_into().unwrap_or_else(|_| unreachable!()))
 ```

### Meta

- What version of Rust GCC were you using: 0d5fe7b9aee

--- a/gcc/rust/parse/rust-parse-impl.h
+++ b/gcc/rust/parse/rust-parse-impl.h
@@ -9055,7 +9055,7 @@ Parser<ManagedTokenSource>::parse_closure_param ()
   AST::AttrVec outer_attrs = parse_outer_attributes ();
 
   // parse pattern (which is required)
-  std::unique_ptr<AST::Pattern> pattern = parse_pattern ();
+  std::unique_ptr<AST::Pattern> pattern = parse_pattern_no_alt ();
   if (pattern == nullptr)
     {
       // not necessarily an error

--- a/gcc/testsuite/rust/compile/closure_no_type_anno.rs
+++ b/gcc/testsuite/rust/compile/closure_no_type_anno.rs
@@ -1,0 +1,5 @@
+// { dg-additional-options "-fsyntax-only" }
+
+pub fn foo() {
+  let a = |_| 15;
+}

